stdenv: mesonFlags: specify more targetPrefix'd binaries

Artturin opened the pull request:
see commit msgs


###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
meson: pull patch to use more binutils variables

meson did not use these

therefore some packages had to do things
like this

```
    substituteInPlace meson.build --replace \
      "find_program('nm')" \
      "find_program('${stdenv.cc.targetPrefix}nm')"
```

alyssais commented:
Can you please explain in the commit message what the purpose of this change is?  "improve" isn't really enough information and even the linked PR doesn't really have the context.

alyssais commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -394,6 +394,20 @@ else let
             [binaries]
             llvm-config = 'llvm-config-native'
+
+            ${lib.optionalString (stdenv.cc ? bintools && stdenv.cc.bintools.targetPrefix != "") ''
```
Do we need to check `stdenv.cc.bintools.targetPrefix != ""`?

Artturin replied:
not really, i just wanted to keep the meson file small but on second thought its better to keep the behaviour the same.

Artturin committed:
treewide: remove unnecessary meson find_program patches

they are unnecessary since 32e36bd429f804fd1468f51e39020f7360d8a5a8
https://github.com/NixOS/nixpkgs/pull/183507

Artturin commented:
TODO: investigate adding using the env vars we already set by adding entries to 
https://github.com/mesonbuild/meson/blob/1e1dee738b7eae9e9c22601522798f4945391379/mesonbuild/envconfig.py#L114

Artturin commented:
https://github.com/mesonbuild/meson/pull/10640

alyssais commented:
The commit message is fantastic now, thank you! :)  (And thanks for going upstream!)

Artturin committed:
stdenv: mesonFlags: use canExecute in needs_exe_wrapper

this is more accurate

SuperSandro2000 commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -29,6 +29,11 @@ python3.pkgs.buildPythonApplication rec {
+    # Use more binutils variables, so we don't have to define them in stdenv.
+    # pr has been merged
+    # https://github.com/mesonbuild/meson/pull/10640
+    ./more-binutils-vars.patch
```
why are we not using fetchpatch?

Artturin replied:
switched

Ericson2314 reviewed (approved):
This is really great! Great job!!

alyssais commented:
This broke `pkgsStatic.glib`:

```
FAILED: glib/tests/getpwuid-preload.so
x86_64-unknown-linux-musl-gcc  -o glib/tests/getpwuid-preload.so glib/tests/getpwuid-preload.so.p/getpwuid-preload.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,getpwuid-preload.so -Wl,--end-group
/nix/store/cja158bhs3cplv0w8zdib822524rwqw9-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/vn93aipraqpva34gz3y9vzjfw3kzdyac-x86_64-unknown-linux-musl-stage-final-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-musl/11.3.0/crtbeginT.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
/nix/store/cja158bhs3cplv0w8zdib822524rwqw9-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
```

alyssais commented:
Ah, because now glib builds its tests, and lots of those tests involve building shared libraries.  Maybe I can get them to disable the tests if default_library=static.
